audit.go

```diff
@@ -505,6 +500,64 @@ func (c *AuditClient) getReply(seq uint32) (*syscall.NetlinkMessage, error) {
 	return &msg, nil
 }
 
+// unset our pid from the audit subsystem and close the socket.
+// This is a sort of isolated refactor, meant to deal with the deadlocks that can happen when we're not careful with blocking operations throughout a lot of this code.
+func (c *AuditClient) closeAndUnsetPid() error {
+	msg := syscall.NetlinkMessage{
+		Header: syscall.NlMsghdr{
+			Type:  AuditSet,
+			Flags: syscall.NLM_F_REQUEST,
+		},
+		Data: AuditStatus{
+			Mask: AuditStatusPID,
+			PID:  0,
+		}.toWireFormat(),
+	}
+
+	noParse := func(bytes []byte) ([]syscall.NetlinkMessage, error) {
+		return nil, nil
+	}
+
+	// If our request to unset the PID would block, then try to drain events from
+	// the netlink socket, resend, try again.
+	// In netlink, EAGAIN usually indicates our read buffer is full.
+	// The auditd code (which I'm using as a reference implementation) doesn't wait for a response when unsetting the audit pid.
+	maxLoop := 5
+	for i := 0; i < maxLoop; i++ {
+		_, err := c.Netlink.SendNoWait(msg)
+		// if we get an interrupt, retry the send
+		if err == nil {
+			return nil
+		} else if errors.Is(err, syscall.EINTR) {
+			// got interrupt, try again
+			continue
+		} else if errors.Is(err, syscall.EAGAIN) {
+			maxRecv := 10000
+			// send would block, try to drain the receive socket
+			for i := 0; i < maxRecv; i++ {
+				_, err = c.Netlink.Receive(true, noParse)
+				if errors.Is(err, syscall.EAGAIN) {
+					// receive would block, try to send again
+					break
+				} else if err == nil || errors.Is(err, syscall.EINTR) || errors.Is(err, syscall.ENOBUFS) {
+					// retry the receive
+					continue
+				} else {
+					// if we have another kind of error, just bail and return that error.
+					return err
+				}
```

Comment:
```suggestion
				switch err {
				case nil, syscall.EINTR, syscall.ENOBUFS:
					// retry the receive
					continue
				case syscall.EAGAIN:
					// receive would block, try to send again
					continue outer
				default:
					// if we have another kind of error, just bail and return that error.
					return err
				}
```

Reply:
Not sure we can cleanly use a switch block if we're still using the `errors.Is` idiom.